### PR TITLE
[ACR] Add ACR Build commands

### PR DESCRIPTION
**Pull request by djyou**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/6241
This is an **experimental preview** for @microsoft users.

**Comment by djyou** [comment]
@derekbekoe Yes we wanted to start the review process while working on publishing the new SDK https://github.com/Azure/azure-sdk-for-python/pull/1935. I also owe tests for the new commands and will add them shortly.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 1]
```diff
@@ -317,3 +327,110 @@
...
             az acr replication update -n MyReplication -r MyRegistry --tags key1=value1 key2=value2
 """
+
+helps['acr build-task create'] = """
+    type: command
+    short-summary: Creates a new build definition which can be triggered by git commits or base image updates.
+    examples:
+        - name: Create a build definition which updates on git commits and base image updates.
+          text: >
+            az acr build-task create -t helloworld:{{.Build.ID}} -n helloworld -r myRegistry -c https://github.com/Azure-Samples/acr-build-helloworld-node --git-access-token $PAT
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: What's `$PAT`?
Would suggest a different example value here?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 2]
```diff
@@ -317,3 +327,110 @@
...
+            az acr build-task delete -n MyBuildTask -r MyRegistry
+"""
+
+helps['acr build-task update'] = """
+    type: command
+    short-summary: Update a build task for a container registry.
+    examples:
+        - name: Update the git access token for a build definition in a container registry.
+          text: >
+            az acr build-task update -n MyBuildTask -r MyRegistry --git-access-token $PAT
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: Same here with `$PAT`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 3]
```diff
@@ -82,8 +95,37 @@ def load_arguments(self, _):
...
-        c.argument('registry_name', options_list=['--registry', '-r'], help='The name of the container registry. You can configure the default registry name using `az configure --defaults acr=<registry name>`', completer=get_resource_name_completion_list(REGISTRY_RESOURCE_TYPE), configured_default='acr')
+        c.argument('registry_name', options_list=['--registry', '-r'])
         c.argument('replication_name', options_list=['--name', '-n'], help='The name of the replication.', completer=get_resource_name_completion_list(REPLICATION_RESOURCE_TYPE))
 
     with self.argument_context('acr replication create') as c:
         c.argument('replication_name', help='The name of the replication. Default to the location name.', completer=None)
+
+    with self.argument_context('acr build') as c:
+        c.argument('registry_name', options_list=['--registry', '-r'])
+        c.positional('source_location', help="The local source code directory path (e.g., './src') or the URL to a git repository (e.g., 'https://github.com/Azure-Samples/acr-build-helloworld-node.git') or a remote tarball (e.g., 'http://server/context.tar.gz').")
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: A FileCompleter may be useful here?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py [hunk 4]
```diff
@@ -248,7 +248,7 @@ def validate_premium_registry(cli_ctx, registry_name, resource_group_name=None,
     """
     registry, resource_group_name = get_registry_by_name(cli_ctx, registry_name, resource_group_name)
 
-    if not registry.sku or registry.sku.name != SkuName.premium.value:
+    if not registry.sku or registry.sku.name != 'Premium':
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_utils.py]: Much prefer using SDK values when possible.
Why this change? Does the SDK no longer document the skus available?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+        print("Sending build context ({0:.3f} {1}) to ACR.".format(size, unit))
+    else:
+        source_location = _check_remote_source_code(source_location)
+        print("Sending build context to ACR.")
+
+    if image_names:
+        is_push_enabled = True
+    else:
+        is_push_enabled = False
+        print("'--image -t' is not provided. Skip image push after build.")
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Use `logger.warning` for these user style messages.

Comment (derekbekoe) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py]: Same with below on messages like 'Waiting for a build agent...' etc.

By using logger, the messages go to stderr instead of stdout.

**Comment by derekbekoe** [comment]
You may want to suppress your current extension as well?
See https://github.com/Azure/azure-cli/blob/e7ca2aefa479829496fa35f17ce7dc36880e473a/doc/authoring_command_modules/authoring_commands.md#extension-suppression

An example here - https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-rdbms/azure/cli/command_modules/rdbms/__init__.py#L24-L29

**Comment by djyou** [comment]
@derekbekoe CI seems to be complaining the use of positional parameters. Could you please advise? We only have one such parameter.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_constants.py [hunk 6]
```diff
@@ -9,6 +9,9 @@
 REGISTRY_RESOURCE_TYPE = ACR_RESOURCE_PROVIDER + '/registries'
 WEBHOOK_RESOURCE_TYPE = REGISTRY_RESOURCE_TYPE + '/webhooks'
 REPLICATION_RESOURCE_TYPE = REGISTRY_RESOURCE_TYPE + '/replications'
+BUILD_TASK_RESOURCE_TYPE = REGISTRY_RESOURCE_TYPE + '/buildTasks'
+BUILD_STEP_RESOURCE_TYPE = BUILD_TASK_RESOURCE_TYPE + '/steps'
 
 CLASSIC_REGISTRY_SKU = ['Classic']
 MANAGED_REGISTRY_SKU = ['Basic', 'Standard', 'Premium']
+BOOL_CHOICES = ['true', 'false']
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_constants.py]: You should not use this.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py [hunk 7]
```diff
@@ -139,3 +183,33 @@ def _get_value(item, *args):
...
+    return 'Manual'
+
+
+def _format_datetime(date_string):
+    from dateutil.parser import parse
+    try:
+        return parse(date_string).strftime("%Y-%m-%dT%H:%M:%SZ")
+    except ValueError:
+        logger.debug("Unable to parse date_string '%s'", date_string)
+        return date_string or ' '
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py]: Why a debug message and potentially unexpected behavior rather than a CLIError?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py [hunk 8]
```diff
@@ -139,3 +183,33 @@ def _get_value(item, *args):
...
+    from dateutil.parser import parse
+    try:
+        duration = parse(finish_time) - parse(start_time)
+        hours = "{0:02d}".format((24 * duration.days) + (duration.seconds // 3600))
+        minutes = "{0:02d}".format((duration.seconds % 3600) // 60)
+        seconds = "{0:02d}".format(duration.seconds % 60)
+        return "{0}:{1}:{2}".format(hours, minutes, seconds)
+    except ValueError:
+        logger.debug("Unable to get duration with start_time '%s' and finish_time '%s'", start_time, finish_time)
+        return ' '
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py]: Same here--if the user inputs an invalid format, then failing silently and having unexpected behavior is probably not better than throwing an error.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 9]
```diff
@@ -82,8 +97,37 @@ def load_arguments(self, _):
...
+        c.argument('registry_name', options_list=['--registry', '-r'])
+        # build task parameters
+        c.argument('build_task_name', options_list=['--name', '-n'], help='The name of the build task.', completer=get_resource_name_completion_list(BUILD_TASK_RESOURCE_TYPE))
+        c.argument('alias', help='The alternative name for build task. Default to the build task name.')
+        c.argument('status', help='The current status of build task.', choices=['Enabled', 'Disabled'])
+        c.argument('os_type', options_list=['--os'], help='The operating system type required for the build.', choices=['Linux', 'Windows'])
+        c.argument('cpu', type=int, help='The CPU configuration in terms of number of cores required for the build.')
+        c.argument('timeout', type=int, help='Build timeout in seconds.')
+        c.argument('repository_url', options_list=['--context', '-c'], help="The full URL to the source code respository.")
+        c.argument('commit_trigger_enabled', help="Indicates whether the source control commit trigger is enabled.", choices=BOOL_CHOICES)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: For anywhere and everywhere "BOOL_CHOICES" is used, see:
https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#registering-flags

**Comment by tjprescott** [comment]
@djyou reviewing the stack trace, it looks like certain stages of the CI load parameters twice, which is why your one positional argument is being found twice and throwing an error. I'll make a PR to fix this.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py (hunk 7 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py]: We have a few cases that the response doesn't have the datetime yet. For example, if users trigger a build and immediately get the status of it before it finishes, we don't have a finish time here. This is table formatting path and we thought we can just show empty in the column rather than throwing an error.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py (hunk 8 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py]: This is the same reason as above handling response datetime. We don't format user inputs here. In this case, `finish_time` could be empty.

Comment (tjprescott) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_format.py]: I didn't notice this was for table format.  Displaying blank is completely appropriate.

**Comment by tjprescott** [comment]
@derekbekoe I recommend we merge this. No idea what's going on with CI and it's recent constant timeouts on Integration.
